src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs

```diff
@@ -0,0 +1,230 @@
+
+            _statement = statement;
+            _modelInfoProvider = modelInfoProvider;
+            _searchParameterDefinitionManager = searchParameterDefinitionManager;
+        }
+
+        public static ICapabilityStatementBuilder Create(IModelInfoProvider modelInfoProvider, ISearchParameterDefinitionManager searchParameterDefinitionManager)
+        {
+            EnsureArg.IsNotNull(modelInfoProvider, nameof(modelInfoProvider));
+            EnsureArg.IsNotNull(searchParameterDefinitionManager, nameof(searchParameterDefinitionManager));
+
+            string manifestName = $"{typeof(CapabilityStatementBuilder).Namespace}.{modelInfoProvider.Version}.BaseCapabilities.json";
+
+            using (Stream resourceStream = Assembly.GetExecutingAssembly().GetManifestResourceStream(manifestName))
+            using (var reader = new StreamReader(resourceStream))
+            {
+                var statement = JsonConvert.DeserializeObject<ListedCapabilityStatement>(reader.ReadToEnd());
+
+                return new CapabilityStatementBuilder(statement, modelInfoProvider, searchParameterDefinitionManager);
+            }
+        }
+
+        public ICapabilityStatementBuilder Update(Action<ListedCapabilityStatement> action)
+        {
+            EnsureArg.IsNotNull(action, nameof(action));
+
+            action(_statement);
+
+            return this;
+        }
+
+        public ICapabilityStatementBuilder UpdateRestResourceComponent(string resourceType, Action<ListedResourceComponent> action)
+        {
+            EnsureArg.IsNotNullOrEmpty(resourceType, nameof(resourceType));
+            EnsureArg.IsNotNull(action, nameof(action));
+            EnsureArg.IsTrue(_modelInfoProvider.IsKnownResource(resourceType), nameof(resourceType), x => GenerateTypeErrorMessage(x, resourceType));
+
+            ListedRestComponent listedRestComponent = _statement.Rest.Server();
+            ListedResourceComponent resourceComponent = listedRestComponent.Resource.SingleOrDefault(x => string.Equals(x.Type, resourceType, StringComparison.OrdinalIgnoreCase));
+
+            if (resourceComponent == null)
+            {
+                resourceComponent = new ListedResourceComponent
+                {
+                    Type = resourceType,
+                };
+                listedRestComponent.Resource.Add(resourceComponent);
+            }
+
+            action(resourceComponent);
+
+            return this;
+        }
+
+        public ICapabilityStatementBuilder TryAddRestInteraction(string resourceType, string interaction)
+        {
+            EnsureArg.IsNotNullOrEmpty(resourceType, nameof(resourceType));
+            EnsureArg.IsNotNullOrEmpty(interaction, nameof(interaction));
+            EnsureArg.IsTrue(_modelInfoProvider.IsKnownResource(resourceType), nameof(resourceType), x => GenerateTypeErrorMessage(x, resourceType));
+
+            UpdateRestResourceComponent(resourceType, c =>
```

Comment:
Is the capability statement built before the server starts? (e.g., does this need to be thread-safe?)

Reply:
We lock inside `SystemConformanceProvider`